src/rgw/rgw_asio_frontend.cc

```diff
@@ -65,67 +62,103 @@ void Pauser::wait()
 using tcp = boost::asio::ip::tcp;
 namespace beast = boost::beast;
 
-// coroutine to handle a client connection to completion
-static void handle_connection(RGWProcessEnv& env, tcp::socket socket,
-                              boost::asio::yield_context yield)
-{
-  auto cct = env.store->ctx();
-  boost::system::error_code ec;
+class Connection {
+  RGWProcessEnv& env;
+  boost::asio::strand strand;
+  tcp::socket socket;
 
-  // limit header to 4k, since we read it all into a single buffer
-  constexpr size_t header_limit = 4096;
+  // references are bound to callbacks for async operations. if a callback
+  // function returns without issuing another operation, the reference is
+  // dropped and the Connection is deleted/closed
+  std::atomic<int> nref{0};
+  using Ref = boost::intrusive_ptr<Connection>;
+
+  // limit header to 4k, since we read it all into a single flat_buffer
+  static constexpr size_t header_limit = 4096;
   // don't impose a limit on the body, since we read it in pieces
-  constexpr size_t body_limit = std::numeric_limits<size_t>::max();
+  static constexpr size_t body_limit = std::numeric_limits<size_t>::max();
 
   beast::flat_buffer buffer{header_limit};
```

Comment:
If you are using `parser::header_limit` you can just default construct this flat buffer.